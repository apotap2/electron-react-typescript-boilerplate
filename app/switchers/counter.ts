import { ActionSwitcher, createActionFactory } from "action_switcher";

export type TState = number;

export const switcher = new ActionSwitcher<TState>(0);

export const increment = createActionFactory(switcher, {
  apply(state: TState): TState {
      return state + 1;
  },
  TYPE: "INCREMENT_COUNTER",
});

export const decrement = createActionFactory(switcher, {
  apply(state: TState): TState {
      return state - 1;
  },
  TYPE: "DECREMENT_COUNTER",
});

export function incrementIfOdd() {
  return (dispatch: Function, getState: Function) => {
    const { counter } = getState();

    if (counter % 2 === 0) {
      return;
    }

    dispatch(increment());
  };
}

export function incrementAsync(delay: number = 1000) {
  return (dispatch: Function) => {
    setTimeout(() => {
      dispatch(increment());
    }, delay);
  };
}

export const Actions = {
  increment,
  decrement,
  incrementIfOdd,
  incrementAsync,
};
