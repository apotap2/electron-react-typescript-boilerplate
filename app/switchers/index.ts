import { combineReducers, Reducer } from 'redux';
import { routerReducer as routing } from 'react-router-redux';
import { switcher as CounterSwitcher, TState as TCounterState } from './counter';

const rootReducer = combineReducers({
  counter: (state, action) => {
    if (state === undefined) {
      return CounterSwitcher.getInitialState();
    }
    return CounterSwitcher.apply(state, action);
  },
  routing: routing as Reducer<any>
});

export interface IState {
  counter: TCounterState;
}

export default rootReducer;
