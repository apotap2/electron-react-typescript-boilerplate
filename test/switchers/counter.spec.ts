import { spy } from 'sinon';
import { switcher, Actions } from '../../app/switchers/counter';
import { increment, decrement } from '../../app/switchers/counter';
import { IAction } from 'action_switcher';

const counter = (state: number, action: IAction) => {return switcher.apply(state, action);};

describe('switchers', () => {
  describe('actions', () => {
    it('should increment should create increment action', () => {
      expect(Actions.increment()).toMatchSnapshot();
    });
  
    it('should decrement should create decrement action', () => {
      expect(Actions.decrement()).toMatchSnapshot();
    });
  });

  describe('async actions', () => {
    it('should incrementIfOdd should create increment action', () => {
    const fn = Actions.incrementIfOdd();
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    const getState = () => ({ counter: 1 });
    fn(dispatch, getState);
    expect(dispatch.calledWith({ type: Actions.increment.TYPE() })).toBe(true);
  });

  it('should incrementIfOdd shouldnt create increment action if counter is even', () => {
    const fn = Actions.incrementIfOdd();
    const dispatch = spy();
    const getState = () => ({ counter: 2 });
    fn(dispatch, getState);
    expect(dispatch.called).toBe(false);
  });

  // There's no nice way to test this at the moment...
  it('should incrementAsync', done => {
    const fn = Actions.incrementAsync(1);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    setTimeout(() => {
      expect(dispatch.calledWith({ type: Actions.increment.TYPE() })).toBe(true);
      done();
    }, 5);
  });
  });

  describe('reducer', () => {
    it('should handle INCREMENT_COUNTER', () => {
      expect(counter(1, increment())).toBe(2);
    });

    it('should handle DECREMENT_COUNTER', () => {
      expect(counter(1, decrement())).toBe(0);
    });

    it('should handle unknown action type', () => {
      expect(counter(1, { type: 'unknown' })).toBe(1);
    });
  });
  
});
