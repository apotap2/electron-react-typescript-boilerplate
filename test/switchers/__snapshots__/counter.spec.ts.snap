// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`switchers actions should decrement should create decrement action 1`] = `
Object {
  "type": "DECREMENT_COUNTER",
}
`;

exports[`switchers actions should increment should create increment action 1`] = `
Object {
  "type": "INCREMENT_COUNTER",
}
`;
